### PR TITLE
Add performance calculations

### PR DESCRIPTION
**Pull request by LavermanJJ**
Add performance calculations to the heat pump:
* cop heating
* cop cooling
* overall performance
* overall performance heating
* overall performance drinking water 

I've introduced an abstract class `Part` to ensure `DataValue(Part)` and `PerformanceCalculator(Part)` have the `scaled_value` property required for the HA-integration. 

I've added the calculations to the heat pump component, and did not create a new component, as these calculations belong only to the heat pump and do not affect other components.

### CODE REVIEW

**Comment by LavermanJJ** [comment]
@LLukas22 review, as always, appreciated. 👍

File: pysolarfocus/components/base/component.py [hunk 1]
```diff
@@ -86,7 +89,11 @@ def has_input_address(self)->bool:
     @property
     def has_holding_address(self)->bool:
         return self.holding_address >= 0 and self.holding_count > 0
-      
+    
+    @property
+    def has_performance_calculators(self)->bool:
+        return  self.__get_performance_calculators()
```
Comment (LLukas22) [inline pysolarfocus/components/base/component.py:95]: Maybe add a length check here, atm you are returning a list as a boolean

File: pysolarfocus/components/base/component.py [hunk 2]
```diff
@@ -113,6 +120,14 @@ def __get_holding_values(self)->list[tuple[str,DataValue]]:
             self.__holding_values = sorted(self.__holding_values,key=lambda item: item[1].address)
         return self.__holding_values
     
+    def __get_performance_calculators(self)->list[tuple[str,PerformanceCalculator]]:
```
Comment (LLukas22) [inline pysolarfocus/components/base/component.py:123]: This should probably be called in the initialize method, atm there is no obvious way to initialize this without calling `has_performance_calculators`

File: pysolarfocus/components/base/part.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from abc import abstractmethod
+
+
+class Part(object):
+    """
+    Performing performance calculations
```
Comment (LLukas22) [inline pysolarfocus/components/base/part.py:6]: Smart, but the comment is kind of misleading.
Maybe change it to something like "abstraction of a metric of the heating system".

File: pysolarfocus/components/base/performance_calculator.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        self.nominator = nominator
+        self.denominator = denominator
+    
+    @property
+    def scaled_value(self)->float:
+        """
+        To provide a uniform interface for integrating, a scaled value property is offered
+        """
+        if self.denominator.scaled_value:
+            self.value =  self.nominator.scaled_value / self.denominator.scaled_value
```
Comment (LLukas22) [inline pysolarfocus/components/base/performance_calculator.py:22]: If `scaled_value` is never called `self.value` is never assigned and in an unknown state. Maybe assign None to it in the `__init__`

File: pysolarfocus/components/heat_pump.py [hunk 5]
```diff
@@ -29,28 +30,8 @@ def __init__(self) -> None:
...
-    def performance_overall_drinking_water(self) -> float:
-        if self.electrical_energy_drinking_water.scaled_value:
-            return self.thermal_energy_drinking_water.scaled_value / self.electrical_energy_drinking_water.scaled_value
-        return 0.0
-
+        self.cop_heating = PerformanceCalculator(self.thermal_power_heating, self.electrical_power)
+        self.cop_cooling = PerformanceCalculator(self.thermal_power_cooling, self.electrical_power)
+        self.performance_overall = PerformanceCalculator(self.thermal_energy_total, self.electrical_energy_total)
+        self.performance_overall_heating = PerformanceCalculator(self.thermal_energy_heating, self.electrical_energy_heating)
+        self.performance_overall_drinking_water = PerformanceCalculator(self.thermal_energy_drinking_water, self.electrical_energy_drinking_water)
```
Comment (LLukas22) [inline pysolarfocus/components/heat_pump.py:37]: Looks a lot cleaner than the explicit properties 👍

Comment (LavermanJJ) [reply pysolarfocus/components/heat_pump.py:37]: in deed. 😄

**Pull request merged**